File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # 7,2,5,1,6,4
        l, h = 0, 0
        res = float("-inf")
        while h < len(prices):
            res = max(res, prices[h] - prices[l])
            h += 1
            if h<len(prices) and prices[h] < prices[l]:
                l = h
        return res

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // 7,1,5,3,6,4
        //     ^
        int currMin = prices[0];
        int profit = 0;
        for (int price : prices) {
            if (currMin <= price) {
                profit = Math.max(profit, price - currMin);
            } else {
                currMin = price;
            }
        }
        return profit;
    }
}